README.md
# Nepeta

Plugin for VCV Rack

## Module List

- BitMixer
- PolyXform
- Rotor

## Manual Installation

Copy the `Nepeta-<version>-<OS>.vcvplugin` for your OS from [release](https://github.com/Letheward/Nepeta/releases) to your VCV Rack Plugin Directory (you can find it in Rack by clicking Help - Open user folder, and go into `plugins`).

## How to Use

See [`docs/Manual.md`](docs/Manual.md).

The `docs` directory is included in the distribution so you can view it offline.   
(in `Nepeta` of Plugin Directory after Rack decompressed the `.vcvplugin`)

## How to Build

Follow the [official instruction](https://vcvrack.com/manual/PluginDevelopmentTutorial) from VCV to build. (todo: need testing) 

However, I personally don't use the official setup (packages from MSYS2).  
I use a pre-bundled MinGW64 and manually installed every program needed to run the build system.

## Licenses and Credits

Code are licensed under [MIT License](https://opensource.org/licenses/MIT).  

Graphic design and documentation are licensed under [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/).  

Font used: [Montserrat](https://github.com/JulietaUla/Montserrat)



docs/Changelog.md
# Changelog

## 2.0.0-alpha1

- New module **BitMixer**
- New module **PolyXform**
- New module **Rotor**

src/Rotor.cpp
#include "plugin.hpp"



/* ==== Math ==== */

struct Vector3 {float     x,  y,  z;};
struct Rotor3D {float s, yz, zx, xy;};

// R* v R
Vector3 v3_rotate(Vector3 v, Rotor3D r) {

    // temp result, a vector and a trivector
    float _x   =  v.x * r.s  - v.y * r.xy + v.z * r.zx;
    float _y   =  v.y * r.s  - v.z * r.yz + v.x * r.xy;
    float _z   =  v.z * r.s  - v.x * r.zx + v.y * r.yz;
    float _xyz = -v.x * r.yz - v.y * r.zx - v.z * r.xy;

    // trivector in result will always be 0
    return (Vector3) {
        _x * r.s - _y * r.xy + _z * r.zx - _xyz * r.yz,
        _y * r.s - _z * r.yz + _x * r.xy - _xyz * r.zx,
        _z * r.s - _x * r.zx + _y * r.yz - _xyz * r.xy
    };
}

Rotor3D r3d_mul(Rotor3D a, Rotor3D b) {
    return (Rotor3D) {
        a.s * b.s  - a.yz * b.yz - a.zx * b.zx - a.xy * b.xy,
        a.s * b.yz + a.yz * b.s  - a.zx * b.xy + a.xy * b.zx,
        a.s * b.zx + a.zx * b.s  - a.xy * b.yz + a.yz * b.xy,
        a.s * b.xy + a.xy * b.s  - a.yz * b.zx + a.zx * b.yz
    };
}

Rotor3D r3d_normalize(Rotor3D r) {
    float l = sqrtf(r.s * r.s + r.yz * r.yz + r.zx * r.zx + r.xy * r.xy);
    if (l == 0) return (Rotor3D) {1, 0, 0, 0};
    return (Rotor3D) {r.s / l, r.yz / l, r.zx / l, r.xy / l};
}




/* ==== Module ==== */

struct RotorModule: Module {

    enum ParamId {
        ROTOR_S,
        ROTOR_YZ,
        ROTOR_ZX,
        ROTOR_XY,
        PARAMS_LEN
    };

    enum InputId {
        ROTOR_INPUT_S,
        ROTOR_INPUT_YZ,
        ROTOR_INPUT_ZX,
        ROTOR_INPUT_XY,
        INPUT_X,
        INPUT_Y,
        INPUT_Z,
        INPUTS_LEN
    };

    enum OutputId {
        OUTPUT_X,
        OUTPUT_Y,
        OUTPUT_Z,
        OUTPUTS_LEN
    };

    enum LightId {
        LIGHTS_LEN
    };
    
    int input_mode = 0;

    RotorModule() {

        config(PARAMS_LEN, INPUTS_LEN, OUTPUTS_LEN, LIGHTS_LEN);

        configParam(ROTOR_S,  -1, 1, 0,  "s");
        configParam(ROTOR_YZ, -1, 1, 0, "yz");
        configParam(ROTOR_ZX, -1, 1, 0, "zx");
        configParam(ROTOR_XY, -1, 1, 0, "xy");

        configInput(ROTOR_INPUT_S,   "s");
        configInput(ROTOR_INPUT_YZ, "yz");
        configInput(ROTOR_INPUT_ZX, "zx");
        configInput(ROTOR_INPUT_XY, "xy");

        configInput(INPUT_X, "x");
        configInput(INPUT_Y, "y");
        configInput(INPUT_Z, "z");

        configOutput(OUTPUT_X, "x");
        configOutput(OUTPUT_Y, "y");
        configOutput(OUTPUT_Z, "z");
    }

    void process(const ProcessArgs& args) override {
        
        float offset = (float) input_mode;

        Rotor3D rotor_input = {
            inputs[ROTOR_INPUT_S ].isConnected() ? (inputs[ROTOR_INPUT_S ].getVoltage() / 5.0f - offset) : 0.0f,
            inputs[ROTOR_INPUT_YZ].isConnected() ? (inputs[ROTOR_INPUT_YZ].getVoltage() / 5.0f - offset) : 0.0f,
            inputs[ROTOR_INPUT_ZX].isConnected() ? (inputs[ROTOR_INPUT_ZX].getVoltage() / 5.0f - offset) : 0.0f,
            inputs[ROTOR_INPUT_XY].isConnected() ? (inputs[ROTOR_INPUT_XY].getVoltage() / 5.0f - offset) : 0.0f,
        };

        Rotor3D rotor_param = {
            params[ROTOR_S ].getValue(),
            params[ROTOR_YZ].getValue(),
            params[ROTOR_ZX].getValue(),
            params[ROTOR_XY].getValue(),
        };

        Vector3 vector_input = {
            inputs[INPUT_X].getVoltage(),
            inputs[INPUT_Y].getVoltage(),
            inputs[INPUT_Z].getVoltage(),
        };

        Rotor3D rotor = r3d_mul(r3d_normalize(rotor_input), r3d_normalize(rotor_param));
        Vector3 out   = v3_rotate(vector_input, rotor);

        outputs[OUTPUT_X].setVoltage(out.x);
        outputs[OUTPUT_Y].setVoltage(out.y);
        outputs[OUTPUT_Z].setVoltage(out.z);
    }
};


struct Rotor_Widget: ModuleWidget {

    Rotor_Widget(RotorModule* module) {

        // we use px here
        // 1HP        = 15px
        // Height: 3U = 380px
        // 6HP = 15px * 6 = 90px

        setModule(module);
        setPanel(createPanel(asset::plugin(plugin_instance, "res/Rotor.svg")));

        // "C++ Template is so great!!! It reduces code duplication!!!"
        #define make_param(name, widget, x, y)  addParam(createParamCentered<widget>(Vec(x, y), module, RotorModule::name))
        #define make_input(name, widget, x, y)  addInput(createInputCentered<widget>(Vec(x, y), module, RotorModule::name))
        #define make_output(name, widget, x, y) addOutput(createOutputCentered<widget>(Vec(x, y), module, RotorModule::name))

        make_input(ROTOR_INPUT_S,  Nepeta_Port,      25, 380 * 0.15);
        make_input(ROTOR_INPUT_YZ, Nepeta_Port,      65, 380 * 0.15);
        make_input(ROTOR_INPUT_ZX, Nepeta_Port,      25, 380 * 0.25);
        make_input(ROTOR_INPUT_XY, Nepeta_Port,      65, 380 * 0.25);

        make_param(ROTOR_S,        Nepeta_SmallKnob, 25, 380 * 0.35);
        make_param(ROTOR_YZ,       Nepeta_SmallKnob, 65, 380 * 0.35);
        make_param(ROTOR_ZX,       Nepeta_SmallKnob, 25, 380 * 0.45);
        make_param(ROTOR_XY,       Nepeta_SmallKnob, 65, 380 * 0.45);

        make_input(  INPUT_X,      Nepeta_Port,      25, 380 * 0.65);
        make_input(  INPUT_Y,      Nepeta_Port,      25, 380 * 0.75);
        make_input(  INPUT_Z,      Nepeta_Port,      25, 380 * 0.85);
        make_output(OUTPUT_X,      Nepeta_Port,      65, 380 * 0.65);
        make_output(OUTPUT_Y,      Nepeta_Port,      65, 380 * 0.75);
        make_output(OUTPUT_Z,      Nepeta_Port,      65, 380 * 0.85);

        #undef make_param
        #undef make_input
        #undef make_output
    }

    void appendContextMenu(Menu* menu) override {

        RotorModule* module = dynamic_cast<RotorModule*> (this->module);
        assert(module);

        menu->addChild(new MenuSeparator);
        menu->addChild(createIndexSubmenuItem(
            "Rotor Input Mode",
            {"Bipolar", "Unipolar"},
            [=]() {return module->input_mode;},
            [=](int i) {module->input_mode = i;}
        ));
    }
};

Model* model_Rotor = createModel<RotorModule, Rotor_Widget>("Rotor");



src/plugin.hpp
/* ==== Dealing With API and C++ ==== */

#pragma once
#include <rack.hpp>

using namespace rack;

// Declare the Plugin, defined in plugin.cpp
extern Plugin* plugin_instance; 

// Declare each Model, defined in each module source file
extern Model* model_BitMixer;
extern Model* model_PolyXform;
extern Model* model_Rotor;




/* ==== Common Widgets ==== */

// todo: we don't use namespace rack here, find out if this is bad

struct Nepeta_SmallKnob: app::SvgKnob {
    Nepeta_SmallKnob() {
        minAngle = -0.8 * M_PI;
        maxAngle =  0.8 * M_PI;
        setSvg(Svg::load(asset::plugin(plugin_instance, "res/Widgets/SmallKnob.svg")));
    }
};

struct Nepeta_BigKnob: app::SvgKnob {
    Nepeta_BigKnob() {
        minAngle = -0.75 * M_PI;
        maxAngle =  0.75 * M_PI;
        setSvg(Svg::load(asset::plugin(plugin_instance, "res/Widgets/BigKnob.svg")));
    }
};

struct Nepeta_Port: app::SvgPort {
    Nepeta_Port() {
        setSvg(Svg::load(asset::plugin(plugin_instance, "res/Widgets/Port.svg")));
    }
};



src/PolyXform.cpp
#include "plugin.hpp"



/* ==== Math ==== */

/*
note: matrices are store in row major order here

memory
------>
*  *  *  *
*  *  *  *
*  *  *  *
*  *  *  *

A         B                out
m=3 n=2   n=2 p=4          m=3 p=4
*  *                       *  *  *  *
*  *      *  *  *  *       *  *  *  *
*  *      *  *  *  *       *  *  *  *

*/

// out = AB, like in math, means first B then A
static inline void matrix_mul(float* out, float* A, float* B, int m, int n, int p) {
    for (int i = 0; i < m; i++) {
        for (int k = 0; k < n; k++) {
            for (int j = 0; j < p; j++) {
                out[i * p + j] += A[i * n + k] * B[k * p + j];
            }
        }
    }
}







/* ==== Module ==== */

struct PolyXform: Module {

    enum ParamId {
        MATRIX1_ROW,
        MATRIX1_COL,
        MATRIX2_ROW,
        MATRIX2_COL,
        SCALE1,
        SCALE2,
        PARAMS_LEN
    };

    enum InputId {
        POLY_INPUT1,
        POLY_INPUT2,
        INPUTS_LEN
    };

    enum OutputId {
        OUTPUT,
        OUTPUTS_LEN
    };

    enum LightId {
        ENUMS(MATRIX_LIGHTS, 16 * 2),
        LIGHTS_LEN
    };

    PolyXform() {

        config(PARAMS_LEN, INPUTS_LEN, OUTPUTS_LEN, LIGHTS_LEN);
        configParam(MATRIX1_ROW,   1, 4, 2, "Matrix 1 Row");
        configParam(MATRIX1_COL,   1, 4, 2, "Matrix 1 Column");
        configParam(MATRIX2_ROW,   1, 4, 2, "Matrix 2 Row");
        configParam(MATRIX2_COL,   1, 4, 2, "Matrix 2 Column");
        configParam(SCALE1,        0, 1, 0.5, "Scale 1");
        configParam(SCALE2,        0, 1, 0.5, "Scale 2");
        configInput(POLY_INPUT1, "Matrix 1");
        configInput(POLY_INPUT2, "Matrix 2");
        configOutput(OUTPUT,     "Matrix");

        paramQuantities[MATRIX1_ROW]->snapEnabled = true;
        paramQuantities[MATRIX1_COL]->snapEnabled = true;
        paramQuantities[MATRIX2_ROW]->snapEnabled = true;
        paramQuantities[MATRIX2_COL]->snapEnabled = true;

        divider.setDivision(16);
    }

	dsp::ClockDivider divider;

    void process(const ProcessArgs& args) override {
        
        float matrix1   [16] = {0};
        float matrix2   [16] = {0};
        float matrix_out[16] = {0};

        int row1 = (int) params[MATRIX1_ROW].getValue();
        int col1 = (int) params[MATRIX1_COL].getValue();
        int row2 = (int) params[MATRIX2_ROW].getValue();
        int col2 = (int) params[MATRIX2_COL].getValue();

        float scale1 = params[SCALE1].getValue();
        float scale2 = params[SCALE2].getValue();
        
        int min_side = col1 < row2 ? col1 : row2; // get the n in m n p, choose min of those because we want data to be fully filled
		for (int i = 0; i < row1 * min_side; i++)  matrix1[i] = inputs[POLY_INPUT1].getVoltage(i) * scale1;
        for (int i = 0; i < min_side * col2; i++)  matrix2[i] = inputs[POLY_INPUT2].getVoltage(i) * scale2;

        matrix_mul(matrix_out, matrix1, matrix2, row1, min_side, col2);

        int out_count = row1 * col2;
        outputs[OUTPUT].channels = out_count; 
        for (int i = 0; i < out_count; i++) {
            outputs[OUTPUT].setVoltage(matrix_out[i], i);
        }

        /* ---- Set Lights ---- */
		if (!divider.process()) return;

        float dt = args.sampleTime * divider.getDivision();
        for (int i = 0; i < 16 * 2; i++) {
            lights[MATRIX_LIGHTS + i].setSmoothBrightness(0, dt);
        }

        for (int i = 0; i < row1; i++) {
            for (int j = 0; j < col2; j++) {
                float v = matrix_out[i * col2 + j] / 10.0;
                lights[MATRIX_LIGHTS + (i * 4 + j) * 2    ].setSmoothBrightness( v, dt);
                lights[MATRIX_LIGHTS + (i * 4 + j) * 2 + 1].setSmoothBrightness(-v, dt);
            }
        }
    }
};


struct PolyXform_Widget: ModuleWidget {

    PolyXform_Widget(PolyXform* module) {

        // we use px here
        // 1HP        = 15px
        // Height: 3U = 380px
        // 6HP = 15px * 6 = 90px

        setModule(module);
        setPanel(createPanel(asset::plugin(plugin_instance, "res/PolyXform.svg")));

        // "C++ Template is so great!!! It reduces code duplication!!!"
        #define make_param(name, widget, x, y)  addParam(createParamCentered<widget>(Vec(x, y), module, PolyXform::name))
        #define make_input(name, widget, x, y)  addInput(createInputCentered<widget>(Vec(x, y), module, PolyXform::name))
        #define make_output(name, widget, x, y) addOutput(createOutputCentered<widget>(Vec(x, y), module, PolyXform::name))

        make_param(MATRIX1_ROW, Nepeta_SmallKnob, 25, 380 * 0.35);
        make_param(MATRIX1_COL, Nepeta_SmallKnob, 65, 380 * 0.35);
        make_param(MATRIX2_ROW, Nepeta_SmallKnob, 25, 380 * 0.45);
        make_param(MATRIX2_COL, Nepeta_SmallKnob, 65, 380 * 0.45);

        make_input(POLY_INPUT1, Nepeta_Port,      25, 380 * 0.6);
        make_param(SCALE1,      Nepeta_SmallKnob, 65, 380 * 0.6);

        make_input(POLY_INPUT2, Nepeta_Port,      25, 380 * 0.7);
        make_param(SCALE2,      Nepeta_SmallKnob, 65, 380 * 0.7);

        make_output(OUTPUT,     Nepeta_Port,      45, 380 * 0.85);
        
        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < 4; j++) {
                addChild(createLightCentered<MediumSimpleLight<GreenRedLight>>(
                    Vec(22.5 + j * 15, 54 + i * 15), 
                    module, 
                    PolyXform::MATRIX_LIGHTS + 2 * (i * 4 + j)
                ));
            }
        }

        #undef make_param
        #undef make_input
        #undef make_output
    }
};



Model* model_PolyXform = createModel<PolyXform, PolyXform_Widget>("PolyXform");



src/BitMixer.cpp
#include "plugin.hpp"



/* ==== Utils ==== */

typedef unsigned int u32;

static inline u32 bit_shift(u32 in, int n) {
    if (n >= 0) return in >>  n;
    else        return in << -n;
}

// maybe use asm instruction?
static inline u32 bit_rotate(u32 in, int n) {
    if (n == 0) return in; // shift by 32 (see below) is UB so we need a 0 check, also maybe faster?
    if (n >  0) return (in >>  n) | (in << (32 - n));
    else        return (in << -n) | (in >> (32 + n));
}

static inline float sqrt_clamp(float in) {
    float out = abs(in);
    if (out > 1000000000) return clamp(in, -5.0, 5.0);
    while (out > 5.0) out = sqrtf(out);
    return in > 0 ? out : -out;
}




/* ==== Module ==== */

struct BitMixer: Module {

    enum ParamId {
        MIX_MODE,
        SIGN_MODE,
        GAIN1,
        GAIN2,
        BIT_SHIFT1,
        BIT_SHIFT2,
        OUT_GAIN,
        PARAMS_LEN
    };

    enum InputId {
        DATA_INPUT1,
        DATA_INPUT2,
        INPUTS_LEN
    };

    enum OutputId {
        OUTPUT,
        OUTPUTS_LEN
    };

    enum LightId {
        LIGHTS_LEN
    };

    BitMixer() {

        config(PARAMS_LEN, INPUTS_LEN, OUTPUTS_LEN, LIGHTS_LEN);
        configParam(MIX_MODE,      1, 8, 1, "Main Mix Mode");
        configParam(SIGN_MODE,     0, 5, 0, "Sign Mix Mode");
        configParam(BIT_SHIFT1, -31, 31, 0, "Bit Shift 1");
        configParam(BIT_SHIFT2, -31, 31, 0, "Bit Shift 2");
        configParam(GAIN1,         0, 2, 1, "Gain 1",   " dB", -10, 40);
        configParam(GAIN2,         0, 2, 1, "Gain 2",   " dB", -10, 40);
        configParam(OUT_GAIN,      0, 2, 1, "Out Gain", " dB", -10, 40);
        configInput(DATA_INPUT1, "Data 1");
        configInput(DATA_INPUT2, "Data 2");
        configOutput(OUTPUT,     "Mix");

        paramQuantities[MIX_MODE  ]->snapEnabled = true;
        paramQuantities[SIGN_MODE ]->snapEnabled = true;
        paramQuantities[BIT_SHIFT1]->snapEnabled = true;
        paramQuantities[BIT_SHIFT2]->snapEnabled = true;
    }

    int shift_mode = 0;
    int clip_mode  = 0;

    void process(const ProcessArgs& args) override {

        /* ---- Get Data ---- */
        float out_gain  =       params[OUT_GAIN ].getValue();
        int   mix_mode  = (int) params[MIX_MODE ].getValue();
        int   sign_mode = (int) params[SIGN_MODE].getValue();

        u32 sign1, sign2;
        u32 data1, data2;
        {
            float value1 = inputs[DATA_INPUT1].getVoltage() * params[GAIN1].getValue();
            float value2 = inputs[DATA_INPUT2].getVoltage() * params[GAIN2].getValue();
            int   shift1 = (int) params[BIT_SHIFT1].getValue();
            int   shift2 = (int) params[BIT_SHIFT2].getValue();

            u32* p1 = (u32*) &value1;
            u32* p2 = (u32*) &value2;
            data1 = *p1;
            data2 = *p2;

            sign1 = data1 & 0x80000000;
            sign2 = data2 & 0x80000000;

            if (shift_mode) {
                data1 = bit_shift(data1, shift1);
                data2 = bit_shift(data2, shift2);
            } else {
                data1 = bit_rotate(data1, shift1);
                data2 = bit_rotate(data2, shift2);
            }
        }

        /* ---- Process ---- */
        u32 bits = 0;
        switch (mix_mode) {
            case 1: bits = (data1 > data2) ? data1 : data2; break;
            case 5: bits = data2 ? (data1 / data2) : data1; break;
            case 2: bits = data1 + data2; break;
            case 3: bits = data1 - data2; break;
            case 4: bits = data1 * data2; break;
            case 6: bits = data1 & data2; break;
            case 7: bits = data1 | data2; break;
            case 8: bits = data1 ^ data2; break;
            default: break;
        }

        if (sign_mode) {

            bits = bits << 1 >> 1;
            switch (sign_mode) {
                case 1: bits |= sign1;         break;
                case 2: bits |= sign2;         break;
                case 3: bits |= sign1 & sign2; break;
                case 4: bits |= sign1 | sign2; break;
                case 5: bits |= sign1 ^ sign2; break;
                default:                       break;
            }
        }

        /* ---- Post ---- */
        float out;
        {
            float* p = (float*) &bits;
            out = *p;
        }

        out *= out_gain;
        switch (clip_mode) {
            case 0:  out = clamp(out,  -5.0,  5.0); break;
            case 1:  out = tanhf(out * 0.2) * 5.0;  break; // precision lost?
            case 2:  out = sqrt_clamp(out);         break;
            case 3:  out = clamp(out,   0.0, 10.0); break;
            default: out = clamp(out, -10.0, 10.0); break;
        }
        outputs[OUTPUT].setVoltage(out);
    }
};


struct BitMixer_Widget: ModuleWidget {

    BitMixer_Widget(BitMixer* module) {

        // we use px here
        // 1HP        = 15px
        // Height: 3U = 380px
        // 6HP = 15px * 6 = 90px

        setModule(module);
        setPanel(createPanel(asset::plugin(plugin_instance, "res/BitMixer.svg")));

        // "C++ Template is so great!!! It reduces code duplication!!!"
        #define make_param(name, widget, x, y)  addParam(createParamCentered<widget>(Vec(x, y), module, BitMixer::name))
        #define make_input(name, widget, x, y)  addInput(createInputCentered<widget>(Vec(x, y), module, BitMixer::name))
        #define make_output(name, widget, x, y) addOutput(createOutputCentered<widget>(Vec(x, y), module, BitMixer::name))

        make_param(MIX_MODE,    Nepeta_BigKnob,   45, 380 * 0.2);

        make_param(SIGN_MODE,   Nepeta_SmallKnob, 25, 380 * 0.35);
        make_param(OUT_GAIN,    Nepeta_SmallKnob, 65, 380 * 0.35);
        make_param(BIT_SHIFT1,  Nepeta_SmallKnob, 25, 380 * 0.45);
        make_param(BIT_SHIFT2,  Nepeta_SmallKnob, 65, 380 * 0.45);

        make_input(DATA_INPUT1, Nepeta_Port,      25, 380 * 0.6);
        make_param(GAIN1,       Nepeta_SmallKnob, 65, 380 * 0.6);

        make_input(DATA_INPUT2, Nepeta_Port,      25, 380 * 0.7);
        make_param(GAIN2,       Nepeta_SmallKnob, 65, 380 * 0.7);

        make_output(OUTPUT,     Nepeta_Port,      45, 380 * 0.85);

        #undef make_param
        #undef make_input
        #undef make_output
    }

    void appendContextMenu(Menu* menu) override {

        BitMixer* module = dynamic_cast<BitMixer*> (this->module);
        assert(module);

        menu->addChild(new MenuSeparator);
        menu->addChild(createIndexSubmenuItem(
            "Bit Shift Mode",
            {"rotate", "shift"},
            [=]() {return module->shift_mode;},
            [=](int i) {module->shift_mode = i;}
        ));
        menu->addChild(createIndexSubmenuItem(
            "Clipping Mode",
            {"hard", "tanh", "sqrt", "CV"},
            [=]() {return module->clip_mode;},
            [=](int i) {module->clip_mode = i;}
        ));
    }
};

Model* model_BitMixer = createModel<BitMixer, BitMixer_Widget>("BitMixer");

